src/formats/XYZ.cpp

```diff
@@ -78,19 +103,396 @@ optional<uint64_t> XYZFormat::forward() {
+    for (auto c: s) {
+        if (is_whitespace(c)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+std::string write_extended_comment_line(const Frame& frame) {
+    std::string result;
+    // TODO: store more properties? Atom-level properties are a good idea, but
+    // we don't have guarantee that they exist for all atoms. Residue
+    // information could also be nice (resid/resname).
+    result += "Properties=species:S:1:pos:R:3";
+
+    // support for Lattice
+    if (frame.cell().shape() != UnitCell::INFINITE) {
+        auto matrix = frame.cell().matrix();
+        // set small elements to 0
+        for (size_t i=0; i<3; i++) {
+            for (size_t j=0; j<3; j++) {
+                if (matrix[i][j] < 1e-12) {
+                    matrix[i][j] = 0;
+                }
+            }
+        }
+        result += fmt::format(" Lattice=\"{:g} {:g} {:g} {:g} {:g} {:g} {:g} {:g} {:g}\"",
+            matrix[0][0], matrix[1][0], matrix[2][0],
+            matrix[0][1], matrix[1][1], matrix[2][1],
+            matrix[0][2], matrix[1][2], matrix[2][2]
+        );
+    }
+
+    auto sorted_properties = std::map<string_view, Property>{frame.properties().begin(), frame.properties().end()};
+    // support for generic frame properties
+    for (const auto& it: sorted_properties) {
+        if (contains_space(it.first)) {
+            warning("Extended XYZ", "frame property '{}' contains spaces, skipping", it.first);
+            continue;
+        }
+
+        result += fmt::format(" {}=", it.first);
+        switch (it.second.kind()) {
+        case Property::STRING:
+            result += fmt::format("\"{}\"", it.second.as_string());
+            break;
+        case Property::BOOL:
+            result += fmt::format("{}", it.second.as_bool() ? 'T' : 'F');
+            break;
+        case Property::DOUBLE:
+            result += fmt::format("{:g}", it.second.as_double());
+            break;
+        case Property::VECTOR3D:
+            {
+                auto v = it.second.as_vector3d();
+                result += fmt::format("\"{:g} {:g} {:g}\"", v[0], v[1], v[2]);
+            }
+            break;
+        }
+    }
+    return result;
+}
+
+/// mapping name=>properties values. The use of string_view is safe since the
+/// original string_view (as returned by file.readline()) will stay alive longer
+/// than this map.
+using extended_xyz_properties_map_t = std::unordered_map<string_view, string_view>;
+
+/// A simple parser for the extended XYZ comment line. This parser bails out as
+/// soons as possible if  the line does not seems to follow extended XYZ
+/// convention.
+///
+/// This line should contains key/values pairs separated by spaces, while the
+/// key and the value are separated by an = sign. If the value contains spaces,
+/// it is then quoted, both single quotes and double quotes are accepted here.
+/// Example: Properties=species:S:1:pos:R:3:bad:I:1 Lattice=" 3 0 0 0 4 0 0 0 5" name='simple file'
```

Comment:
Serves the leading space in `Lattice="_3...` any purpose?

Reply:
no, not really